### ISSUE TITLE
Possible memory leak with complex tenant types

### ISSUE THREAD

**Issue by tillig** [saaskit/saaskit#86]
TL;DR: If you use a complex tenant type (the `TTenant` in `TenantContext<TTenant>`) and the tenant instances aren't a static/fixed set of objects or objects that implement `IEquatable<T>`, the caching in `TenantPipelineMiddleware` can create a memory leak.

---

There seem to be two separate concepts - the `Tenant` and the `TenantContext<TTenant>` - which are intertwined but used differently with respect to caching.
- The `MemoryCacheTenantResolver` [caches a `TenantContext<TTenant>`](https://github.com/saaskit/saaskit/blob/1.1.4/src/SaasKit.Multitenancy/MemoryCacheTenantResolver.cs#L70) on a sliding expiration.
- The `TenantPipelineMiddleware` [caches a `RequestDelegate` keyed off `TTenant`](https://github.com/saaskit/saaskit/blob/1.1.4/src/SaasKit.Multitenancy/Internal/TenantPipelineMiddleware.cs#L16) in a non-expiring `ConcurrentDictionary`.

In the `AspNetMvcAuthDemo` no issues are exhibited because the set of `AppTenant` objects (the `TTenant` in `TenantContext<TTenant>`) are [a fixed set of instances](https://github.com/saaskit/saaskit/blob/1.1.4/samples/AspNetMvcAuthDemo/AppTenantResolver.cs#L20).

However, if you look at the `CachingAppTenantResolver` in the `AspNetSample` [the `AppTenant` objects get created as they need to be resolved](https://github.com/saaskit/saaskit/blob/1.1.4/samples/AspNetSample/CachingAppTenantResolver.cs#L48). This can cause the following scenario if you mix it with multitenant middleware:

- First request comes in for Tenant A.
  - `CachingAppTenantResolver` creates a new `AppTenant` object in a new `TenantContext<AppTenant>` and caches that. The `TenantContext<AppTenant>` gets returned.
  - The `TenantPipelineMiddleware` lazily adds an entry to the `ConcurrentDictionary<TTenant, Lazy<RequestDelegate>>` with a new tenant-specific app pipeline.
  - The first request completes.
- Second request immediately comes in for Tenant A.
  - `CachingAppTenantResolver` finds the cached `TenantContext<AppTenant>` and returns it.
  - `TenantPipelineMiddleware` finds the cached `RequestDelegate` and runs it.
  - The second request completes.
- Time passes - enough for the `MemoryCacheTenantResolver` to expire the `TenantContext<AppTenant>` object for Tenant A.
- Third request comes in for Tenant A.
  - `CachingAppTenantResolver` creates a new `AppTenant` object in a new `TenantContext<AppTenant>` and caches that. **This is a different `AppTenant` object than the first one.** The `TenantContext<AppTenant>` gets returned.
  - `TenantPipelineMiddleware` uses object reference equality to locate an `AppTenant` matching the one in effect for this request. It's Tenant A, but that _instance_ of `AppTenant` hasn't been seen yet, so _another copy_ of the `RequestDelegate` gets added to the `ConcurrentDictionary`.
  - The third request completes.

Since the `ConcurrentDictionary<TTenant, Lazy<RequestDelegate>>` is keyed by the `TTenant` it assumes either object reference equality will work - implying the tenant objects are a static/singleton set where that will work - or it's a type that implements `IEquatable<T>` like `string` so actual content comparison will work.

None of the examples as they stand really exhibit this because the only one using multitenant middleware has a static set of `TTenant` objects; the example where the `TTenant` is dynamic doesn't have multitenant middleware.

I'm not sure there's anything to be done about that beyond documentation, but it's a gotcha I ran into. It's almost like the `TTenant` really needs to be `TTenantId` to hint that it's sort of a master ID / cache key and all tenant contextual information (configuration, etc.) really needs to be part of `TenantContext<T>` instead.

**Comment by tillig** [comment]
Additional challenge: If you use a simple type like `string` as the tenant (`TTenant`) to work around some of the memory leak/comparison issues here, you can't use `services.AddMultitenancy<TTenant, TResolver>()` [because it adds the `TTenant` as a service](https://github.com/saaskit/saaskit/blob/1.1.4/src/SaasKit.Multitenancy/MultitenancyServiceCollectionExtensions.cs#L24) and adding `string` as a resolvable service is not so great.

**Comment by tillig** [comment]
Oh, looks like this is a duplicate of #76.
